### DISCUSSION TITLE
Obtain the layer of each element in the PDF

### DISCUSSION THREAD

**Discussion by more-strive** [pymupdf/PyMuPDF#2986] [Looking for help] [answered]
**Is your feature request related to a problem? Please describe.**
I have an image editing application here 
Demo: https://yft.design 
Github: https://github.com/dromara/yft-design
I hope to obtain the layer of elements (image, path, text) when importing PDF, so that the position, content, and size of all elements can be resolved now
```
def get_elements(page):
    draws = page.get_drawings()
    texts = page.get_text('dict', flags=11)
    images = page.get_image_info(hashes=True, xrefs=True)
    return [*draws, *images, *texts]
```

**Describe the solution you'd like**

**Describe alternatives you've considered**
Are there several options for how your request could be met?

**Additional context**
![image](https://github.com/pymupdf/PyMuPDF/assets/113762408/c5bfb230-a9ed-4849-aebd-951375fb6f52)
![image](https://github.com/pymupdf/PyMuPDF/assets/113762408/830fd8ae-3d61-4043-ba82-6e2c58d0f402)

**Comment by JorjMcKie** [comment]
Everything you want is already implemented it seems.
You can extract images with all metadata, same is true for vector graphics and for text.
Please be more specific about what you are missing.

**Comment by more-strive** [comment]
![image](https://githubfast.com/pymupdf/PyMuPDF/assets/113762408/0d3f200a-5af8-4993-9b7d-6d6da8069d32)
The square, trapezoid, and other paths in the middle of the image, The parsed path is covered by an image. Sorry, I didn't mark it
```
def get_elements(page):
    draws = page.get_drawings()
    texts = page.get_text('dict', flags=11)
    images = page.get_image_info(hashes=True, xrefs=True)
    return [*draws, *images, *texts]
```
The code here parses the image element after the path, so is there any way to redefine their index

**Comment by JorjMcKie** [comment]
This is a typical "Discussions" post. Let me move this accordingly.

**Reply by more-strive** [reply]
![image](https://github.com/pymupdf/PyMuPDF/assets/113762408/b72b8f53-ffe5-432c-a0ff-1745ae4e651b)
![image](https://github.com/pymupdf/PyMuPDF/assets/113762408/2528b337-cb73-40cf-9348-f034185c81f4)
The path element has been parsed, but it is covered by an image

**Reply by more-strive** [reply]
![image](https://github.com/pymupdf/PyMuPDF/assets/113762408/536764e8-337a-486a-8fb5-f86ad4e3d154)

**Comment by JorjMcKie** [comment]
This cannot be changed. You could delete the image of course and then re-insert it **before** everything else on the page.

**Reply by more-strive** [reply]
Is there a way to know the index of the image

**Reply by JorjMcKie** [reply, marked as answer]
Yes: method `Page.get_text("dict")` extracts text **_and_** images when using the default `flags`.
The sequence of the extracted image and text blocks are like in the page's `/Contents`.

The full sequence of all boundary boxes of everything on the page is reflected by the list [`page.get_bboxlog()`](https://pymupdf.readthedocs.io/en/latest/functions.html#Page.get_bboxlog). The items in this list look like `(obj-type, bbox)`.
So you can take the bbox of an image or some text and then determine the index in the bboxlog that contains it.

**Reply by more-strive** [reply]
This interface only obtains the indexes of text and image. How should the index of path be compared with text and image

**Comment by JorjMcKie** [comment]
> This interface only obtains the indexes of text and image. How should the index of path be compared with text and image

That is what I was referring to: there is no way extract every object type in one single method.
So you must take the bboxes of text, images, graphics and lookup their positions in the bboxlog list.

**Reply by more-strive** [reply]
Thank you, I'll give it a try
